implementation HTLC Endorsement to Mitigate Channel Jamming

vincenzopalazzo opened the pull request:
This is an implementation of https://github.com/lightning/bolts/pull/1071


I wrote this when I was on a train in Spain, so it needs some
more work for sure, but I am putting this out to check the CI

I am still digesting some parts of the proposal but my work here also if incomplete.

My self-open questions are but this can not be part of this PR:

- We should signal this in some way this Endorsement to plugins maybe in https://docs.corelightning.org/docs/hooks#htlc_accepted or something similar that allow us to change the value of the endorsement
- Make this change experimental (I do not know how this fit inside the forwarding Endorsement if we required that the node should enable it to understand it)

vincenzopalazzo commented:
This is quite minimal, but it includes the essential work needed to start addressing the channel mitigation issue.

Additionally, I believe it should support modifying the value of `htlc_accepted`. However, I'm uncertain if we need a different type of hook to fully implement the channel jamming proposal. We'll address this as it arises in the future.

rustyrussell commented on wire/peer_wire.csv:
```diff
@@ -245,6 +245,8 @@ msgdata,update_add_htlc,cltv_expiry,u32,
 msgdata,update_add_htlc,onion_routing_packet,byte,1366
 tlvtype,update_add_htlc_tlvs,blinding_point,0
 tlvdata,update_add_htlc_tlvs,blinding_point,blinding,point,
+tlvtype,update_add_htlc_tlvs,endorsed,1
+tlvdata,update_add_htlc_tlvs,endorsed,endorsed,byte,
 msgtype,update_fulfill_htlc,130
```
This needs to be added as a patch, since it's not in the bolts, eg. wire/extracted_peer_12_jamming.patch.

vincenzopalazzo replied:
Thanks, done!

rustyrussell commented on lightningd/pay.c:
```diff
@@ -1116,9 +1117,22 @@ send_payment_core(struct lightningd *ld,
+	 *    - SHOULD set `endorsed` to `1`.
+	 *
+	 * We wait that someone else smarted than me will provide the way to
+	 * calculate it for now it is just a placesolder. */
```
placeholder

vincenzopalazzo replied:
Oh, there is no solder 😸

rustyrussell commented on lightningd/pay.c:
```diff
@@ -1116,9 +1117,22 @@ send_payment_core(struct lightningd *ld,
 	}
 
+
+	/* BOLT 2 (0539ad868a263040087d2790684f69fb28d3fa97):
```
use bolt-xxx #2, which is how we do this.  In this case, use -jamming I'd say, instead of a particular commit?

vincenzopalazzo replied:
Ok yeah moving this to `BIP-jamming #4`, but make sense

rustyrussell commented on channeld/channeld_wire.csv:
```diff
@@ -104,6 +104,7 @@ msgdata,channeld_offer_htlc,cltv_expiry,u32,
 msgdata,channeld_offer_htlc,payment_hash,sha256,
 msgdata,channeld_offer_htlc,onion_routing_packet,u8,1366
 msgdata,channeld_offer_htlc,blinding,?pubkey,
+msgdata,channeld_offer_htlc,endorsed,?u8,
```
Can this be ?bool Might make the rest clearer?

vincenzopalazzo replied:
much better yeah

rustyrussell reviewed (commented):
Looks intersesting, but not for this release, since it doesn't actually do any endorsement :)

vincenzopalazzo commented:
>Looks intersesting, but not for this release, since it doesn't actually do any endorsement :)

I agree that it does not make a lot of sense to put this inside this release, not because we do not have any endorsement, but because I do not have tested it and I do not know if we need another hook or the current one is fine.

For endorsement currently, we need to implement the reputation and look at it, so I am planning to share the implementation done by Carla inside the https://github.com/LNOpenMetrics/go-lnmetrics.reporter before implementing it inside the cln builtin plugins

I know that you have an opinion on the way that how the metrics should be, so I will move incremental step before putting an effort into implementing it in C

vincenzopalazzo commented:
Ok this is rebased on master, and I addressed the review by @rustyrussell 

The only thing that it is missing is adding a couple of tests, and it looks like it is hard to use lnprototest here maybe

Could you consider adding it to this release? @cdecker due that it is under and experimental flag?

vincenzopalazzo commented:
Ok i need to rework this, I would like to take another approach
